File: typescript/linked-list/linked-list.ts
/**
 * Represents a doubly linked list data structure.
 *
 * This class provides methods to manipulate a list of elements,
 * allowing efficient addition and removal from both ends, as well as
 * deletion of specific elements. Each element maintains links to
 * the previous and next elements in the sequence.
 *
 * @template T The type of elements stored in the linked list.
 */
export class LinkedList<T> {
  /**
   * @var head The first node in the linked list. Is `null` if the list is empty.
   */
  private head: Link<T>;
  /**
   * @var tail The last node in the linked list. Is `null` if the list is empty.
   */
  private tail: Link<T>;
  /**
   * @var length The current number of elements in the linked list.
   */
  private length: number;

  /**
   * Initializes a new empty LinkedList.
   */
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /**
   * Adds a new element to the end of the linked list.
   *
   * If the list is empty, the new element becomes both the head and the tail.
   *
   * @param element The element to add to the list.
   */
  public push(element: T) {
    if (this.tail === null) {
      this.initWith(element);
    } else {
      this.tail.next = new Node(element, this.tail, null);
      this.tail = this.tail.next;
    }
    this.length++;
  }

  /**
   * Removes and returns the element from the end of the linked list.
   *
   * @returns The element removed from the end, or `undefined` if the list was empty.
   */
  public pop(): T | undefined {
    if (this.tail === null) {
      return undefined;
    }

    const popped = this.tail.element;
    this.length--;

    if (this.tail.prev === null) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = this.tail.prev;
      this.tail.next = null;
    }

    return popped;
  }

  /**
   * Removes and returns the element from the beginning of the linked list.
   *
   * @returns The element removed from the beginning, or `undefined` if the list was empty.
   */
  public shift(): T | undefined {
    if (this.head === null) {
      return undefined;
    }

    const shiffted = this.head.element;
    this.length--;

    if (this.head.next === null) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = this.head.next;
      this.head.prev = null;
    }

    return shiffted;
  }

  /**
   * Adds a new element to the beginning of the linked list.
   *
   * If the list is empty, the new element becomes both the head and the tail.
   *
   * @param element The element to add to the list.
   */
  public unshift(element: T) {
    if (this.head === null) {
      this.initWith(element);
    } else {
      this.head.prev = new Node(element, null, this.head);
      this.head = this.head.prev;
    }
    this.length++;
  }

  /**
   * Deletes the first occurrence of a specific element from the linked list.
   *
   * If the element is found, it's removed and the list length is decremented.
   * If the element is not found, the list remains unchanged.
   *
   * @param element The element to delete from the list.
   */
  public delete(element: T) {
    for (let current = this.head; current !== null; current = current.next) {
      if (current.element === element) {
        if (current.prev === null) {
          this.head = current.next;

        } else {
          current.prev.next = current.next;
        }

        if (current.next === null) {
          this.tail = current.prev;

        } else {
          current.next.prev = current.prev;
        }

        this.length--;
        return;
      }
    }
  }

  /**
   * Returns the current number of elements in the linked list.
   *
   * @returns The count of elements.
   */
  public count(): number {
    return this.length;
  }

  /**
   * Initializes the list with a single element when it is empty.
   * Sets both head and tail to a new node containing the element.
   *
   * @param element The first element to add to the newly initialized list.
   */
  private initWith(element: T) {
    this.head = new Node(element);
    this.tail = this.head;
  }
}

/**
 * Represents a single node within the LinkedList.
 *
 * Each node holds an element and references to the next and previous nodes in the list.
 *
 * @template T The type of the element stored in the node.
 */
class Node<T> {
  /**
   * @var element The data stored in this node.
   */
  public element: T;
  /**
   * @var next A reference to the next node in the list, or `null` if this is the last node.
   */
  public next: Link<T>;
  /**
   * @var prev A reference to the previous node in the list, or `null` if this is the first node.
   */
  public prev: Link<T>;

  /**
   * Creates a new Node instance.
   *
   * @param element The element to store in this node.
   * @param prev An optional reference to the previous node. Defaults to `null`.
   * @param next An optional reference to the next node. Defaults to `null`.
   */
  constructor(element: T, prev: Link<T> = null, next: Link<T> = null) {
    this.element = element;
    this.next = next;
    this.prev = prev;
  }
}

/**
 * Defines a type alias for a link in the LinkedList, which can be a Node or `null`.
 * This is used to indicate optional or absent connections between nodes.
 *
 * @template T The type of elements in the linked list node.
 */
type Link<T> = Node<T> | null;

File: lua/linked-list/linked-list.lua
--- The public interface for a doubly linked list data structure.
--- @class LinkedList
local LinkedList <const> = {}

--- The internal implementation and data storage for a doubly linked list.
--- @class OwnerLinkedList : LinkedList
--- @field _head Node | nil The first node in the list.
--- @field _tail Node | nil The last node in the list.
--- @field _count integer The number of elements in the list.
local OwnerLinkedList <const> = {}

--- Represents a single element (node) in a linked list.
--- @class Node
--- @field value any The value stored in the node.
--- @field prev Node The previous node in the list.
--- @field next Node The next node in the list.
local Node <const> = {}

--- Creates a new empty doubly linked list.
--- @generic T The type of elements to be stored in the list.
--- @return LinkedList<T> -- A new list instance.
function LinkedList.new()
    return OwnerLinkedList.new()
end

--- Adds an element to the end of the list.
--- @generic T
--- @param self LinkedList<T>
--- @param element T The element to add.
function LinkedList.push(self, element)
    local owner = OwnerLinkedList.as_owner(self)

    owner._count = owner._count + 1

    if owner._tail == nil then
        owner._tail = Node.new(element, nil, nil)
        owner._head = owner._tail
    else
        owner._tail.next = Node.new(element, owner._tail, nil)
        owner._tail = owner._tail.next
    end
end

--- Removes and returns the element at the end of the list.
--- @generic T
--- @param self LinkedList<T>
--- @return T -- The element that was at the end of the list.
function LinkedList.pop(self)
    local owner = OwnerLinkedList.as_owner(self)

    if owner._tail == nil then
        error("can't pop from empty list")
    end

    owner._count = owner._count - 1
    local popped = owner._tail.value

    if owner._tail == owner._head then
        owner._tail = nil
        owner._head = nil
    else
        owner._tail = owner._tail.prev
        owner._tail.next = nil
    end

    return popped
end

--- Adds an element to the beginning of the list.
--- @generic T
--- @param self LinkedList<T>
--- @param element T The element to add.
function LinkedList.unshift(self, element)
    local owner = OwnerLinkedList.as_owner(self)

    owner._count = owner._count + 1

    if owner._head == nil then
        owner._head = Node.new(element, nil, nil)
        owner._tail = owner._head
    else
        owner._head.prev = Node.new(element, nil, owner._head)
        owner._head = owner._head.prev
    end
end

--- Removes and returns the element at the beginning of the list.
--- @generic T
--- @param self LinkedList<T>
--- @return T -- The element that was at the beginning of the list.
function LinkedList.shift(self)
    local owner = OwnerLinkedList.as_owner(self)

    if owner._head == nil then
        error("can't shift from empty list")
    end

    owner._count = owner._count - 1
    local unshifted = owner._head.value

    if owner._head == owner._tail then
        owner._head = nil
        owner._tail = nil
    else
        owner._head = owner._head.next
        owner._head.prev = nil
    end

    return unshifted
end

--- Returns the number of elements in the list.
--- @generic T
--- @param self LinkedList<T>
--- @return integer -- The count of elements.
function LinkedList.count(self)
    return OwnerLinkedList.as_owner(self)._count
end

--- Deletes the first occurrence of a given element from the list.
--- @generic T
--- @param self LinkedList<T>
--- @param element T The element to delete.
function LinkedList.delete(self, element)
    local owner = OwnerLinkedList.as_owner(self)

    local current = owner._head
    while current ~= nil do
        if current.value == element then
            if current.prev ~= nil then
                current.prev.next = current.next
            else
                owner._head = current.next
            end
            if current.next ~= nil then
                current.next.prev = current.prev
            else
                owner._tail = current.prev
            end
            owner._count = owner._count - 1
            break
        else
            current = current.next
        end
    end
end

LinkedList.__index = LinkedList

--- The internal constructor for a new OwnerLinkedList instance.
--- @generic T
--- @return OwnerLinkedList<T>
function OwnerLinkedList.new()
    return setmetatable(
        {
            _head = nil,
            _tail = nil,
            _count = 0
        },
        LinkedList
    )
end

--- Performs a type cast to the internal representation of the list.
--- @generic T
--- @param list LinkedList<T>
--- @return OwnerLinkedList<T> -- The internal list object.
function OwnerLinkedList.as_owner(list)
    return list
end

--- Creates a new Node instance.
--- @generic T
--- @param value T The value for the new node.
--- @param prev Node<T> The previous node.
--- @param next Node<T> The next node.
--- @return Node<T> -- The new node instance.
function Node.new(value, prev, next)
    return {
        value = value,
        prev = prev,
        next = next
    }
end

return LinkedList.new